common/dpcpp/slab_hash.hpp

```diff
@@ -88,16 +158,42 @@ template <typename K, typename T, typename Hash> class SlabHashTable {
 
       sycl::group_barrier(_gr);
     }
-
     return _ans;
   }
 
 private:
+  void alloc_node(sycl::device_ptr<SlabNode<std::pair<K, T>>> &src) {
+    lock();
+    if (src == nullptr) {
+      auto tmp = _heap.malloc_node();
+      *tmp = SlabNode<std::pair<K, T>>({_empty, T()}); //!!!!!!
+      src = tmp;
+    }
+    unlock();
+    _iter = src;
```

Comment:
Is this not in critical section intentionally? What happens if other thread locks and changes `src `before the first one updates `_iter`?

Reply:
src in this case is always the tail in the list that I want to continue
if two threads called alloc, then two cases are possible:
1. threads work with different lists and their srcs do not correlate in any way
2. the threads work in the same list and called alloc on the same src. this means that they both rested on the tail and want to increase the list. then one of them will go to allocate src and the second will wait. when the first thread (which passed) exits the lock, and the second one enters, the src will not be changed, since the second thread will see that it is not nullptr


the only problem that may arise here is that suddenly the src = tmp entry will take a long time and some thread outside in the vile will check for nullptr and see only a partially written address (I'm not sure that this problem exists). I have this problem in ishews, I know about how to solve it